File: android/app/src/main/java/com/tastebyte/erp/features/warehouse/WarehouseViewModel.kt
package com.tastebyte.erp.features.warehouse

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.tastebyte.erp.core.network.ApiClient
import com.tastebyte.erp.models.Material
import com.tastebyte.erp.models.StockCountItem
import com.tastebyte.erp.models.StockCountRequest
import com.tastebyte.erp.models.Warehouse
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class WarehouseListState(
    val warehouses: List<Warehouse> = emptyList(),
    val isLoading: Boolean = false,
    val error: String? = null
)

data class StockCountState(
    val warehouses: List<Warehouse> = emptyList(),
    val selectedWarehouse: Warehouse? = null,
    val materials: List<Material> = emptyList(),
    val countEntries: Map<String, String> = emptyMap(),
    val isLoading: Boolean = false,
    val isSubmitting: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null
)

class WarehouseViewModel : ViewModel() {

    private val _listState = MutableStateFlow(WarehouseListState())
    val listState: StateFlow<WarehouseListState> = _listState.asStateFlow()

    private val _stockCountState = MutableStateFlow(StockCountState())
    val stockCountState: StateFlow<StockCountState> = _stockCountState.asStateFlow()

    fun loadWarehouses() {
        viewModelScope.launch {
            _listState.value = WarehouseListState(isLoading = true)
            try {
                val response = ApiClient.getService().listWarehouses()
                if (response.success && response.data != null) {
                    _listState.value = WarehouseListState(
                        warehouses = response.data,
                        isLoading = false
                    )
                } else {
                    _listState.value = WarehouseListState(
                        isLoading = false,
                        error = response.error ?: "Failed to load warehouses"
                    )
                }
            } catch (e: Exception) {
                _listState.value = WarehouseListState(
                    isLoading = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun loadStockCountData() {
        viewModelScope.launch {
            _stockCountState.value = _stockCountState.value.copy(isLoading = true, error = null)
            try {
                val api = ApiClient.getService()
                val warehouseResponse = api.listWarehouses()
                val materialsResponse = api.listMaterials(pageSize = 100)

                _stockCountState.value = _stockCountState.value.copy(
                    warehouses = warehouseResponse.data ?: emptyList(),
                    materials = materialsResponse.data?.items ?: emptyList(),
                    isLoading = false
                )
            } catch (e: Exception) {
                _stockCountState.value = _stockCountState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun selectWarehouse(warehouse: Warehouse) {
        _stockCountState.value = _stockCountState.value.copy(
            selectedWarehouse = warehouse,
            countEntries = emptyMap()
        )
    }

    fun updateCountEntry(materialId: String, quantity: String) {
        val entries = _stockCountState.value.countEntries.toMutableMap()
        entries[materialId] = quantity
        _stockCountState.value = _stockCountState.value.copy(countEntries = entries)
    }

    fun submitStockCount() {
        val state = _stockCountState.value
        val warehouse = state.selectedWarehouse ?: return

        val items = state.countEntries
            .filter { it.value.isNotBlank() }
            .mapNotNull { (materialId, quantity) ->
                quantity.toDoubleOrNull()?.let { qty ->
                    StockCountItem(
                        materialId = materialId,
                        countedQuantity = qty
                    )
                }
            }

        if (items.isEmpty()) {
            _stockCountState.value = state.copy(error = "Enter at least one quantity")
            return
        }

        viewModelScope.launch {
            _stockCountState.value = state.copy(isSubmitting = true, error = null)
            try {
                val response = ApiClient.getService().submitStockCount(
                    StockCountRequest(
                        warehouseId = warehouse.id,
                        items = items
                    )
                )
                if (response.success) {
                    _stockCountState.value = _stockCountState.value.copy(
                        isSubmitting = false,
                        countEntries = emptyMap(),
                        successMessage = "Stock count submitted successfully"
                    )
                } else {
                    _stockCountState.value = _stockCountState.value.copy(
                        isSubmitting = false,
                        error = response.error ?: "Submission failed"
                    )
                }
            } catch (e: Exception) {
                _stockCountState.value = _stockCountState.value.copy(
                    isSubmitting = false,
                    error = e.message ?: "Network error"
                )
            }
        }
    }

    fun clearSuccessMessage() {
        _stockCountState.value = _stockCountState.value.copy(successMessage = null)
    }
}

File: ios/TasteByteERP/Features/Warehouse/WarehouseViewModel.swift
import Foundation

@MainActor
final class WarehouseViewModel: ObservableObject {
    @Published var warehouses: [Warehouse] = []
    @Published var stockCounts: [StockCount] = []
    @Published var materials: [Material] = []
    @Published var isLoading = false
    @Published var isSubmitting = false
    @Published var errorMessage: String?
    @Published var successMessage: String?
    @Published var selectedWarehouseId: String?

    func loadWarehouses() async {
        isLoading = true
        errorMessage = nil

        do {
            let response: APIResponse<PaginatedResponse<Warehouse>> = try await APIClient.shared.getPaginated(
                APIEndpoints.warehouses,
                perPage: 50
            )
            if response.success, let paginated = response.data {
                warehouses = paginated.items
                if selectedWarehouseId == nil, let first = warehouses.first {
                    selectedWarehouseId = first.id
                }
            } else {
                errorMessage = response.error ?? "Failed to load warehouses"
            }
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to load warehouses"
        }

        isLoading = false
    }

    func loadMaterials() async {
        do {
            let response: APIResponse<PaginatedResponse<Material>> = try await APIClient.shared.getPaginated(
                APIEndpoints.materials,
                perPage: 100
            )
            if response.success, let paginated = response.data {
                materials = paginated.items
            }
        } catch {
            // Supplementary data
        }
    }

    func loadStockCounts(warehouseId: String) async {
        do {
            let params = [URLQueryItem(name: "warehouse_id", value: warehouseId)]
            let response: APIResponse<PaginatedResponse<StockCount>> = try await APIClient.shared.getPaginated(
                APIEndpoints.stockCounts,
                perPage: 100,
                extraParams: params
            )
            if response.success, let paginated = response.data {
                stockCounts = paginated.items
            }
        } catch {
            stockCounts = []
        }
    }

    func submitStockCount(warehouseId: String, materialId: String, countedQuantity: Double) async {
        isSubmitting = true
        errorMessage = nil
        successMessage = nil

        do {
            let body = CreateStockCount(
                warehouseId: warehouseId,
                materialId: materialId,
                countedQuantity: countedQuantity
            )
            let response: APIResponse<StockCount> = try await APIClient.shared.post(
                APIEndpoints.stockCounts,
                body: body
            )
            if response.success {
                successMessage = "Stock count submitted successfully!"
                if let id = selectedWarehouseId {
                    await loadStockCounts(warehouseId: id)
                }
            } else {
                errorMessage = response.error ?? "Failed to submit stock count"
            }
        } catch let error as APIError {
            errorMessage = error.errorDescription
        } catch {
            errorMessage = "Failed to submit stock count"
        }

        isSubmitting = false
    }

    func materialName(for materialId: String) -> String {
        materials.first { $0.id == materialId }?.name ?? "Unknown Material"
    }

    func warehouseName(for warehouseId: String) -> String {
        warehouses.first { $0.id == warehouseId }?.name ?? "Unknown Warehouse"
    }
}
